File: backend/java/src/test/java/de/iits/elocalculatorbackend/match/controller/MatchControllerIntegrationTest.java
package de.iits.elocalculatorbackend.match.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import de.iits.elocalculatorbackend.match.model.dto.MatchRequestDto;
import de.iits.elocalculatorbackend.match.model.entity.Match;
import de.iits.elocalculatorbackend.match.model.enumeration.Outcome;
import de.iits.elocalculatorbackend.match.repository.MatchRepository;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;

import java.util.Comparator;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc
class MatchControllerIntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private MatchRepository matchRepository;

    @Nested
    class GetAllMatches {
        @Test
        void returnAllMatchesFromDatabase() throws Exception {
            String expectedMatchAsJson = objectMapper.writeValueAsString(matchRepository.findAll());
            ResultActions requestResponse = mockMvc.perform(get("/matches"));

            requestResponse.andExpectAll(
                    status().isOk(),
                    content().json(expectedMatchAsJson)
            );
        }
    }

    @Nested
    class CreateMatch {
        @Test
        void doNotSendMatchForCreation() throws Exception {
            ResultActions requestResponse = mockMvc.perform(post("/matches"));

            requestResponse.andExpectAll(
                    status().isBadRequest(),
                    status().reason("Match required for creation, but no match was found in request body")
            );
        }

        @Test
        void sendMatchForCreation() throws Exception {
            MatchRequestDto newMatchResult = createNewMatchResult();
            String newMatchResultAsJson = objectMapper.writeValueAsString(newMatchResult);
            ResultActions requestResponse = mockMvc.perform(post("/matches")
                    .contentType("application/json")
                    .content(newMatchResultAsJson));

            List<Match> allMatches = matchRepository.findAll();
            allMatches.sort(Comparator.comparing(Match::getPlayedOn));
            Match lastMatch = allMatches.get(allMatches.size() - 1);
            assertThat(lastMatch.getId()).isNotNull();
            assertThat(lastMatch.getWhitePlayerUsername()).isEqualTo(newMatchResult.whitePlayerUsername());
            assertThat(lastMatch.getBlackPlayerUsername()).isEqualTo(newMatchResult.blackPlayerUsername());
            assertThat(lastMatch.getOutcome()).isEqualTo(newMatchResult.outcome());
            assertThat(newMatchResult.toEntity().getPlayedOn()).isAfterOrEqualTo(lastMatch.getPlayedOn());

            requestResponse.andExpectAll(status().isOk());
            Match responseBody = objectMapper.readValue(requestResponse.andReturn().getResponse().getContentAsString(), Match.class);
            assertThat(responseBody.getId()).isNotNull();
            assertThat(responseBody.getWhitePlayerUsername()).isEqualTo(newMatchResult.whitePlayerUsername());
            assertThat(responseBody.getBlackPlayerUsername()).isEqualTo(newMatchResult.blackPlayerUsername());
            assertThat(responseBody.getOutcome()).isEqualTo(newMatchResult.outcome());
            assertThat(newMatchResult.toEntity().getPlayedOn()).isAfterOrEqualTo(responseBody.getPlayedOn());
        }

        MatchRequestDto createNewMatchResult() {
            String whitePlayerUsername = "Maxi";
            String blackPlayerUsername = "Lisa";
            Outcome outcome = Outcome.DRAW;
            return new MatchRequestDto(whitePlayerUsername, blackPlayerUsername, outcome);
        }
    }
}

File: backend/springboot-kotlin/src/test/kotlin/de/iits/elo/match/controller/MatchControllerIntegrationTest.kt
package de.iits.elo.match.controller

import com.fasterxml.jackson.databind.ObjectMapper
import de.iits.elo.match.model.dto.MatchRequestDto
import de.iits.elo.match.model.dto.MatchResponseDto
import de.iits.elo.match.model.dto.toEntity
import de.iits.elo.match.model.entity.Match
import de.iits.elo.match.model.enumeration.Outcome
import de.iits.elo.match.repository.MatchRepository
import io.kotest.matchers.date.shouldNotBeAfter
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc
class MatchControllerIntegrationTest {

    @Autowired
    lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var objectMapper: ObjectMapper

    @Autowired
    lateinit var matchRepository: MatchRepository

    @Nested
    inner class GetAllMatches {

        @Test
        fun `return all matches from database`() {
            val expectedMatchAsJson = objectMapper.writeValueAsString(matchRepository.findAll())
            val requestResponse = mockMvc.get("/matches")

            requestResponse.andExpectAll {
                status { isOk() }
                content { json(expectedMatchAsJson) }
            }
        }
    }

    @Nested
    inner class CreateMatch {

        @Test
        fun `do not send match for creation`() {
            val requestResponse = mockMvc.post("/matches")

            requestResponse.andExpectAll {
                status { isBadRequest() }
                status { reason("Match required for creation, but no match was found in request body") }
            }
        }

        @Test
        fun `send valid match for creation`() {
            val newMatchRequest = createNewMatchRequest()
            val newMatchRequestAsJson = objectMapper.writeValueAsString(newMatchRequest)
            val requestResponse = mockMvc.post("/matches") {
                contentType = MediaType.APPLICATION_JSON
                content = newMatchRequestAsJson
            }

            val allMatches = matchRepository.findAll()
            allMatches.sortBy(Match::playedOn)
            allMatches.last().id shouldNotBe null
            allMatches.last().whitePlayerUsername shouldBe newMatchRequest.whitePlayerUsername
            allMatches.last().blackPlayerUsername shouldBe newMatchRequest.blackPlayerUsername
            allMatches.last().outcome shouldBe newMatchRequest.outcome
            allMatches.last().playedOn shouldNotBeAfter newMatchRequest.toEntity().playedOn

            requestResponse.andExpectAll { status { isOk() } }
            val responseBody = objectMapper.readValue(requestResponse.andReturn().response.contentAsString, MatchResponseDto::class.java)
            responseBody.id shouldNotBe null
            responseBody.whitePlayerUsername shouldBe newMatchRequest.whitePlayerUsername
            responseBody.blackPlayerUsername shouldBe newMatchRequest.blackPlayerUsername
            responseBody.outcome shouldBe newMatchRequest.outcome
            responseBody.playedOn shouldNotBeAfter newMatchRequest.toEntity().playedOn
        }

        private fun createNewMatchRequest() =
            MatchRequestDto(
                whitePlayerUsername = "Maxi",
                blackPlayerUsername = "Lisa",
                outcome = Outcome.DRAW,
            )
    }
}
